doc/srcspan.md

```diff
@@ -0,0 +1,124 @@
+The bad thing is that to find out the exact srcspan, we need an extra lookup at array "mix" file.
+
+### b. DWARF method
+
+Nothing peculiar to discuss about this method, and it looks like we already make use of it elsewhere in the plugin (for coverage):
+<https://github.com/input-output-hk/plutus/blob/master/plutus-tx-plugin/src/PlutusTx/Compiler/Types.hs#L80-L88>
+
+### c. Prof method
+
+Pros: The annotations can be controlled by the user by prepending user-interested haskell expressions in the original file
+with "{-# SCC -#}" comments.
+
+Cons: Cabal will refuse to mix the compilation of libraries compiled with -prof and libraries without it, and will try to rebuild those that miss -prof.
+
+### d. Ghci method
+
+I haven't found a way yet to **programmatically** get the `SrcSpan`s out of `ghci`.
+
+Also we seem to rely in our compilation on the flag `-fobject-code` and that may be a problem perhaps for ghci.
+
+### Comparing the detailing of annotations
+
+Some annotate more GHC_Core AST node places than others.
+In the order of most `SrcSpan` annotations generated in the GHC Core node:
+
+`hpc > dwarf > prof = ghci`
+
+, meaning hpc annotated the most places.
+
+### Comparing unfoldings
+
+Ghci does not generate any interface `.hi` files (only keeps relevant data in memory), so
+that means it is going to be difficult to get the unfoldings in a later step and pass them to our `plutus-tx-plugin`.
+
+All the other 3 methods are confirmed to include the `SrcSpan`s annotations inside the unfoldings section of the `.hi` interface files.
+
+### Decision
+
+Go with DWARF `-g` since it is good enough and we used it before.
+
+## 2. Propagating SrcSpans
+
+Our plugin can already accept `CoreExpr` annotated with `SrcSpan`s but strips them off / ignores them.
+
+The change is simply a matter of copying these `SrcSpan` data when constructing the PIR AST.
+And so forth until we reach the final UPLC.
+
+### Note by Michael
+
+When copying the `SrcSpan`s we should use our own `SrcSpan'` datatype, isomoprhic to GHC's https://hackage.haskell.org/package/ghc-8.2.2/docs/SrcLoc.html#t:SrcSpan
+, to constrain the dependency on the `ghc` library only for the `plutus-tx-plugin` package.
+
+### Note about missing SrcSpans
+
+None of the above SrcSpan-collecting methods will annotate every GHC_Core node with a `SrcSpan`;
+some child nodes are implied to belong inside the parent `SrcSpan`.
+
+I don't think we have to do something about it; we may perhaps have
+to make make our plutus (pir/plc/uplc) annotation to be `Maybe SrcSpan` to cover missing srcspans.
+
+Another way is to place an early `CoreExpr->CoreExpr` traversal
+that propagates missing srcspans to child nodes even before we start our pipeline
+(idea taken from `https://hackage.haskell.org/package/ghc-srcspan-plugin`).
+
+A missing srcspan will also not be a problem at runtime (during debugging),
+because the debugger can use the cek `Context` (callstack) to approximate the current position.
```

Comment:
Can you elaborate on this? Do you mean using the SrcSpan associated with the top frame of the stack, if exists?

Reply:
I think the problem is if we e.g. inline stuff. Consider:
```
let f = \x -> 1 <1:1> -- suppose the annotation is on the lambda, anyway not on the 1
in f 2 <2:1>

==>

(\x -> 1) 2 <2:1>

==>

1 <2:1>
```
i.e. one something gets inlined it will appear in a new context, so we can't rely on the enclosing being accurate wrt the original context of the program. So I think we just need to annotate everything!